src/sumo_backend/mysql.clj

```diff
@@ -0,0 +1,207 @@
+;; and scaling connections/threads accordingly.
+
+
+(defn in?
+  "true if collection contains elm"
+  [coll elm]
+  (some #(= elm %) coll))
+
+;; Note on ranks
+;; Komusubi, Maegashira, and Juryo are not the same
+;; Difference between them is the Maegashira's number
+;; Or all Maegashira numbers + the Juryo number
+;; e.g. Maegashira #1 is one rank away from Komusubi
+
+(def ranks
+  { :yokozuna 1
+    :ozeki 2
+    :sekiwake 3
+    :komusubi 4
+    :maegashira 4 ; Maegashira #1, Maegashira #2, ...
+    :juryo 4 })   ; Juryo #1, Juryo #2, ...
+
+(defn get-all-ranks-in-tournament
+  "returns all ranks competing in a tournament"
+  [month year] ; hash set of every rank
+  (into #{} ; how can this be done in one sql call?
+    (concat
+      (map
+        #(:east_rank %)
+        (jdbc/query mysql-db (sql/format
+          (sql/build
+            :select :east_rank
+            :modifiers [:distinct]
+            :from :bout
+            :where
+            [:and
+              [:= :year year]
+              [:= :month month]]))))
+      (map
+        #(:west_rank %)
+        (jdbc/query mysql-db (sql/format
+          (sql/build
+            :select :west_rank
+            :modifiers [:distinct]
+            :from :bout
+            :where
+            [:and
+            [:= :year year]
+            [:= :month month]])))))))
+
+(defn get-rikishi
+  "gets rikihsi record specified by passed in name"
+  [name]
+  (jdbc/query mysql-db (sql/format
+    (sql/build
+      :select :*
+      :from :rikishi
+      :where [:= :name name]))))
+  
+(defn list-rikishi
+  "list all rikishi records"
+  []
+  (jdbc/query mysql-db (sql/format 
+    (sql/build 
+      :select :* 
+      :from :rikishi))))
+
+(defn valid-rikishi?
+  "true if data exists for
+   passed in rikishi string, false otherwise"
+  [rikishi]
+  (let [rikishi-names (map #(get % :name) (list-rikishi))]
+   (if (in? rikishi-names (clojure.string/upper-case rikishi))
+    true
+    false)))
+
+(defn list-bouts
+  "list all bouts data exists for"
+  []
+  (jdbc/query mysql-db (sql/format
+    (sql/build
+      :select [:month :year]
+      :modifiers [:distinct]
+      :from :bout
+      :order-by [[:year :desc] [:month :desc]]))))
+
+(defn get-bouts-by-date
+  "gets all bouts in specified time frame: year, month, day"
+  ([year]
+    (jdbc/query mysql-db (sql/format 
+      (sql/build
+        :select :*
+        :from :bout
+        :where [:= :year year]))))
+  ([year month]
+    (jdbc/query mysql-db (sql/format
+      (sql/build 
+        :select :* 
+        :from :bout
+        :where 
+          [:and
+            [:= :year year]
+            [:= :month month]]))))
+  ([year month day]
+    (jdbc/query mysql-db (sql/format
+      (sql/build
+        :select :*
+        :from :bout
+        :where 
+          [:and
+            [:= :year year]
+            [:= :month month] 
+            [:= :day day]])))))
+
+(defn get-bouts-by-rikishi
```

Comment:
You can simplify this with a self-referential function:
```
(defn get-bouts-by-rikishi
  ([name]
    (jdbc/query
      mysql-db
      (sql/format
        (sql/build
          :select :*
          :from :bout 
          :where [:or [:= :east name] [:= :west name]]))))
  ([name year month day]
    (jdbc/query
      mysql-db
      (sql/format
        (sql/build
          :select :*
          :from :bout
          :where (concat 
                   [:and
                    [:or [:= :east name] [:= :west name]]
                    [[:= year year]]]
                   (when month
                     [[:= month month]])
                   (when day
                     [[:= day day]]))))))
  ([name year month]
    (get-bouts-by-rikishi year month nil))
  ([name year]
    (get-bouts-by-rikishi year nil nil)))

```


This could also be simplified using optional keyword args, which would also give more flexibility (give me day 1 results from all tournaments, etc) something like:
```
(defn get-bouts-by-rikishi
  [name & {:keys [year month day]}]
  (jdbc/query
    mysql-db
    (sql/format
      (sql/build
        :select :*
        :from :bout
        :where
          (let [name-clause [:or [:= :east name] [:= :west name]]]
            (if (or year month day)
              (concat
                [:and name-clause]
                (when year [[:= year year]])
                (when month [[:= month month]])
                (when day [[:= day day]]))
              name-clause))))))
```

Reply:
thank you, updated to use optional keyword args 👍🏼